LFS-80: Add cache busting for the frontend code

OrdiNeu opened the pull request:
This is achieved in three parts:
1) Using the [WebpackAssetsManifest](https://github.com/webdeveric/webpack-assets-manifest) plugin to generate a .json file telling the server where to find the entry points, e.g.:
`{"jcr:primaryType":"sling:Folder","jcr:createdBy":"admin","jcr:created":"Thu Jul 11 2019 15:10:27 GMT-0400","themeindex.js":"themeindex.202e3c3516601f18eb8f.js"}`
2) Switching the homepage from html.GET.html to html.GET.esp to parse out the location of the script.
3) Using [WebpackCleanupPlugin](https://www.npmjs.com/package/webpack-cleanup-plugin) to clean up the webpack folder between compiles, for organization's sake. 

This is all based on suggestions by @sdumitriu through [this Medium article](https://medium.com/@jsilvax/introducing-webpack-into-an-existing-java-based-web-app-ff53f14d37ec).

OrdiNeu committed:
LFS-80: Add cache busting for the frontend code
Add cachebusting using a manifest.json file, and clean up the webpack
folder between compilations.

sdumitriu reviewed (commented):
Can you do the same for the other webpacked files?

OrdiNeu commented on modules/homepage/src/main/frontend/webpack.config.js:
```diff
@@ -1,8 +1,17 @@
+const { CleanWebpackPlugin } = require('clean-webpack-plugin');
+const WebpackAssetsManifest = require('webpack-assets-manifest');
+
```
One provides the plugin as the default export, the other provides it as a named export. I've tried removing the curly braces around CleanWebpackPlugin, but since it's not the default export it fails.

OrdiNeu committed:
LFS-80: Add cache busting for the frontend code
Add cache busting for every webpack module currently in dev.
Undo the change from the homepage .esp file back to an .html file.

OrdiNeu commented:
> Can you do the same for the other webpacked files?

Done

OrdiNeu committed:
LFS-80: Add cache busting for the frontend code
Bugfix on header.js not loading the cache busted file

sdumitriu committed:
LFS-80: Add cache busting for the frontend code
Move all asset descriptions in a single file

OrdiNeu commented on modules/login/src/main/resources/SLING-INF/content/libs/lfs/login/html.html:
```diff
@@ -34,6 +34,8 @@
   <body>
     <div id="main-login-container"></div>
-    <script src="/libs/lfs/resources/login.js"></script>
+    <sly data-sly-use.loginlib="/libs/lfs/resources/login">
```
Oh, interesting -- I had that since `#main-login-container` was overwritten by the scripts in login.js it wouldn't work. Are there any side-effects to the `<script src="/libs/lfs/resources/login.hash.js"></script>` line disappearing from the elements?

sdumitriu replied:
Actually that's not a very valid comment, since the div isn't surrounding, it's on top. A `data-sly-use` attribute is only valid in the scope of the element where it's used. You could, in theory, put the script tags inside the div, but that doesn't make that much sense.

OrdiNeu commented on modules/commons/src/main/frontend/src/header.js:
```diff
@@ -71,8 +71,10 @@ class GlobalHeader extends React.Component {
+      var loginLib = window.Sling.getContent("/libs/lfs/resources/login.json", 1, "");
       const script = document.createElement("script");
-      script.src = "/libs/lfs/resources/loginDialogue.js";
+      script.src = "/libs/lfs/resources/" + loginLib["loginDialogue.js"];
```
Refactored into a login dialog extension? While we can definitely move the `dialogue-login-container` div into its own extension, the trade-off is that in every web page with a header we'd then require two lines of input statements, namely:
```
<sly data-sly-resource="${@selectors='header'}"/>
<sly data-sly-resource="${@selectors='loginDialogue'}"/>
```

sdumitriu replied:
No, an extension should be able to plug itself in without needing any other code. But this is something for the future, not related to this PR.

sdumitriu commented on modules/commons/src/main/frontend/webpack.config.js:
```diff
@@ -1,9 +1,18 @@
+  plugins: [
+    new CleanWebpackPlugin(),
+    new WebpackAssetsManifest({
+      output: "commons.json"
```
What do you think of this?

OrdiNeu replied:
It's brilliant -- I had no idea Sling would merge multiple .json files with the same name into one node. I think it works really well, and unless multiple extensions name the same output file I don't see any problems with it (it'd be unlikely for the hashes to collide, but this might be something to consider. Maybe if we prepended the output filenames with the name of the module?).

sdumitriu replied:
> Maybe if we prepended the output filenames with the name of the module?

Can the assets plugin be configured that way?

Sling doesn't actually merge the JSON files, but the way the content importer works is:

- for every node that is to be imported,
- does its `Sling-Initial-Content` entry in the manifest has `overwrite:=true`?
- if yes, delete the existing node with all its properties and subchildren from the repo if it exists, then create a new one
- otherwise, get the existing node and use it
- for every property defined in the source, add it to the node
- for every child node, repeat the process

So effectively as long as none of our properties overlap, and the standard properties (`jcr:primaryType`) aren't conflicting, it looks like a merge.

OrdiNeu replied:
> Can the assets plugin be configured that way?
You sure can, see my commit. It looks a little strange because it's not something built-in, but it's not terribly difficult.

Re: the content importer: interesting. We'll have to be careful with the `overwrite:=true` lines, but this should work.

OrdiNeu committed:
LFS-80: Add cache busting for the frontend code
Prepend asset output filenames with the name of the module

OrdiNeu committed:
LFS-80: Add cache busting for the frontend code
Alter the names of the .js files in addition to their cachebusted
variants, in order to prevent collisions between modules.
